ocfl-java-aws/src/main/java/edu/wisc/library/ocfl/aws/OcflS3Client.java

```diff
@@ -400,15 +285,22 @@ public Path downloadFile(String srcPath, Path dstPath) {
     @Override
     public InputStream downloadStream(String srcPath) {
         var srcKey = keyBuilder.buildFromPath(srcPath);
-        LOG.debug("Streaming bucket {} key {}", bucket, srcKey);
+        LOG.debug("Streaming from bucket {} key {}", bucket, srcKey);
 
         try {
-            return s3Client.getObject(GetObjectRequest.builder()
-                    .bucket(bucket)
-                    .key(srcKey.getKey())
-                    .build());
-        } catch (NoSuchKeyException e) {
-            throw new KeyNotFoundException(String.format("Key %s not found in bucket %s.", srcKey, bucket), e);
+            return s3Client.getObject(
+                            GetObjectRequest.builder()
+                                    .bucket(bucket)
+                                    .key(srcKey.getKey())
+                                    .build(),
+                            AsyncResponseTransformer.toBlockingInputStream())
+                    .join();
+        } catch (RuntimeException e) {
+            var cause = unwrapCompletionEx(e);
+            if (cause instanceof NoSuchKeyException) {
+                throw new KeyNotFoundException("Key " + srcKey + " not found in bucket " + bucket, cause);
+            }
+            throw new OcflS3Exception("Failed to download " + srcKey, cause);
```

Comment:
I'm getting this error when I try to build an `OcflRepository` with `.ocflConfig(config -> config.setOcflVersion(OcflVersion.OCFL_1_1)`. The error message is `Failed to download .../0=ocfl_1.0`. Should the srcKey end in `0=ocfl_1.1` when the config sets the `OcflVersion` to `OCFL_1_1`?

Reply:
Can you provide the complete stacktrace?

Reply:
<details><summary>stacktrace</summary>
<p>

```
: Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [edu.wisc.library.ocfl.api.OcflRepository]: Factory method 'localCloudOcflRepository' threw exception; nested exception is edu.wisc.library.ocfl.aws.OcflS3Exception: Failed to download preservation-ocfl-test/0=ocfl_1.0
: at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:185)
: at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:653)
: ... 69 common frames omitted
: Caused by: edu.wisc.library.ocfl.aws.OcflS3Exception: Failed to download preservation-ocfl-test/0=ocfl_1.0
: at edu.wisc.library.ocfl.aws.OcflS3Client.downloadStream(OcflS3Client.java:303)
: at edu.wisc.library.ocfl.core.storage.cloud.CloudStorage.read(CloudStorage.java:148)
: at edu.wisc.library.ocfl.core.storage.cloud.CloudStorage.readToString(CloudStorage.java:159)
: at edu.wisc.library.ocfl.core.storage.DefaultOcflStorageInitializer.identifyExistingVersion(DefaultOcflStorageInitializer.java:130)
: at edu.wisc.library.ocfl.core.storage.DefaultOcflStorageInitializer.loadAndValidateExistingRepo(DefaultOcflStorageInitializer.java:102)
: at edu.wisc.library.ocfl.core.storage.DefaultOcflStorageInitializer.initializeStorage(DefaultOcflStorageInitializer.java:88)
: at edu.wisc.library.ocfl.core.storage.DefaultOcflStorage.doInitialize(DefaultOcflStorage.java:626)
: at edu.wisc.library.ocfl.core.storage.AbstractOcflStorage.initializeStorage(AbstractOcflStorage.java:63)
: at edu.wisc.library.ocfl.core.storage.ObjectDetailsDbOcflStorage.doInitialize(ObjectDetailsDbOcflStorage.java:65)
: at edu.wisc.library.ocfl.core.storage.AbstractOcflStorage.initializeStorage(AbstractOcflStorage.java:63)
: at edu.wisc.library.ocfl.core.storage.CachingOcflStorage.doInitialize(CachingOcflStorage.java:59)
: at edu.wisc.library.ocfl.core.storage.AbstractOcflStorage.initializeStorage(AbstractOcflStorage.java:63)
: at edu.wisc.library.ocfl.core.OcflRepositoryBuilder.buildInternal(OcflRepositoryBuilder.java:405)
: at edu.wisc.library.ocfl.core.OcflRepositoryBuilder.build(OcflRepositoryBuilder.java:384)
: at edu.wisc.library.sdg.preservation.worker.config.AppConfig.localCloudOcflRepository(AppConfig.java:384)
: at edu.wisc.library.sdg.preservation.worker.config.AppConfig$$EnhancerBySpringCGLIB$$bf6dbfe.CGLIB$localCloudOcflRepository$26(<generated>)
: at edu.wisc.library.sdg.preservation.worker.config.AppConfig$$EnhancerBySpringCGLIB$$bf6dbfe$$FastClassBySpringCGLIB$$d99d16ca.invoke(<generated>)
: at org.springframework.cglib.proxy.MethodProxy.invokeSuper(MethodProxy.java:244)
: at org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.intercept(ConfigurationClassEnhancer.java:331)
: at edu.wisc.library.sdg.preservation.worker.config.AppConfig$$EnhancerBySpringCGLIB$$bf6dbfe.localCloudOcflRepository(<generated>)
: at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
: at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
: at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
: at java.base/java.lang.reflect.Method.invoke(Method.java:568)
: at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:154)
: ... 70 common frames omitted
: Caused by: software.amazon.awssdk.services.s3.model.S3Exception: null (Service: S3, Status Code: 404, Request ID: 40095382-d48c-4323-908f-a6d6ac08ab0b)
: at software.amazon.awssdk.services.s3.model.S3Exception$BuilderImpl.build(S3Exception.java:104)
: at software.amazon.awssdk.services.s3.model.S3Exception$BuilderImpl.build(S3Exception.java:58)
: at software.amazon.awssdk.protocols.query.internal.unmarshall.AwsXmlErrorUnmarshaller.unmarshall(AwsXmlErrorUnmarshaller.java:99)
: at software.amazon.awssdk.protocols.query.unmarshall.AwsXmlErrorProtocolUnmarshaller.handle(AwsXmlErrorProtocolUnmarshaller.java:102)
: at software.amazon.awssdk.protocols.query.unmarshall.AwsXmlErrorProtocolUnmarshaller.handle(AwsXmlErrorProtocolUnmarshaller.java:82)
: at software.amazon.awssdk.core.http.MetricCollectingHttpResponseHandler.lambda$handle$0(MetricCollectingHttpResponseHandler.java:52)
: at software.amazon.awssdk.core.internal.util.MetricUtils.measureDurationUnsafe(MetricUtils.java:63)
: at software.amazon.awssdk.core.http.MetricCollectingHttpResponseHandler.handle(MetricCollectingHttpResponseHandler.java:52)
: at software.amazon.awssdk.core.internal.http.async.AsyncResponseHandler.lambda$prepare$0(AsyncResponseHandler.java:89)
: at java.base/java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1150)
: at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
: at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2147)
: at software.amazon.awssdk.core.internal.http.async.AsyncResponseHandler$BaosSubscriber.onComplete(AsyncResponseHandler.java:132)
: at software.amazon.awssdk.utils.async.SimplePublisher.doProcessQueue(SimplePublisher.java:275)
: at software.amazon.awssdk.utils.async.SimplePublisher.processEventQueue(SimplePublisher.java:224)
: at software.amazon.awssdk.utils.async.SimplePublisher.complete(SimplePublisher.java:157)
: at java.base/java.util.concurrent.CompletableFuture.uniRunNow(CompletableFuture.java:819)
: at java.base/java.util.concurrent.CompletableFuture.uniRunStage(CompletableFuture.java:803)
: at java.base/java.util.concurrent.CompletableFuture.thenRun(CompletableFuture.java:2195)
: at software.amazon.awssdk.services.s3.internal.crt.S3CrtResponseHandlerAdapter.onErrorResponseComplete(S3CrtResponseHandlerAdapter.java:135)
: at software.amazon.awssdk.services.s3.internal.crt.S3CrtResponseHandlerAdapter.handleError(S3CrtResponseHandlerAdapter.java:124)
: at software.amazon.awssdk.services.s3.internal.crt.S3CrtResponseHandlerAdapter.onFinished(S3CrtResponseHandlerAdapter.java:93)
: at software.amazon.awssdk.crt.s3.S3MetaRequestResponseHandlerNativeAdapter.onFinished(S3MetaRequestResponseHandlerNativeAdapter.java:24)
```

</p>
</details>

Reply:
Thanks, @BeckyAndresen, that was helpful. I just pushed a commit that should address the problem.

When client initializes, it attempts to detect whatever the repository says it is by reading the namaste files in the repo root. This is why it's looking for the 1.0 file, even if you're actually using 1.1. Anyway, the actual issue appears to be that the new s3 client was returning a generic `S3Exception` rather than the `NoSuchKeyException` the code was expecting.

Reply:
These changes fixed the `BeanInstantiationException` issue.

I am now running into a different issue, and will add a new comment once I have more details.

Reply:
@BeckyAndresen : have you been able to surface more details?

Reply:
@awoods -- Thanks for following up. I ran the test (https://github.com/pwinckles/ocfl-java-s3-tester) and sent the logs to @pwinckles today. The test failed with a 403, which is the same issue I had in January.

Reply:
@awoods I suspect that the problem Becky was experiencing may be something system related and a bug with the CRT client, though this has yet to be confirmed. However, I can say for sure that UW-Madison's S3 does appear to support domain-based access.

I took a look at how Harvard is currently using the TransferManager, and it is not currently using the new CRT client. It would be helpful if you all would be able to run the test project that I made (https://github.com/pwinckles/ocfl-java-s3-tester) as an easy way to determine if it would work for you.

If it works for you, I would be inclined to proceed with this PR regardless of whether or not the S3 writes get sorted out for UW-Madison in the near term, because I have a number of additional changes that I want to make that are blocked by this PR, and I believe the reason that the client is not currently working for UW-Madison is a bug with the client rather than an incompatibility with their S3 implementation.

Reply:
Thanks for the details and updates, @BeckyAndresen and @pwinckles . 
I (or the team) will look at this as soon as we can.